File: apps/engine/qresearch/cpp/greeks.cpp
#include <cmath>
#include <limits>
#include "q_pricing.hpp"

namespace pricing {

namespace {
constexpr double kInvSqrt2 = 0.70710678118654752440;
constexpr double kInvSqrt2Pi = 0.39894228040143267794;

double normal_cdf(double x) {
    return 0.5 * std::erfc(-x * kInvSqrt2);
}

double normal_pdf(double x) {
    return kInvSqrt2Pi * std::exp(-0.5 * x * x);
}
}  // namespace

Greeks black_scholes_greeks(
    double spot,
    double strike,
    double time_to_maturity,
    double risk_free_rate,
    double volatility,
    int option_type
) {
    const double nan = std::numeric_limits<double>::quiet_NaN();
    if (!std::isfinite(spot) || !std::isfinite(strike) || !std::isfinite(time_to_maturity)
        || !std::isfinite(risk_free_rate) || !std::isfinite(volatility)
        || spot <= 0 || strike <= 0 || time_to_maturity < 0 || volatility < 0
        || (option_type != 1 && option_type != -1)) return {nan,nan,nan,nan,nan};
    if (time_to_maturity == 0.0 || volatility == 0.0) {
        const double discounted = strike * std::exp(-risk_free_rate*time_to_maturity);
        const double boundary = spot - discounted;
        const bool kink = boundary == 0;
        const double active = option_type*boundary > 0 ? 1 : option_type*boundary < 0 ? 0 : .5;
        return {option_type*active, kink ? nan : 0, kink ? nan : 0,
                time_to_maturity == 0 || kink ? nan : -option_type*active*risk_free_rate*discounted,
                kink && time_to_maturity > 0 ? nan : option_type*active*time_to_maturity*discounted};
    }

    const double sqrt_t = std::sqrt(time_to_maturity);
    const double variance = volatility * volatility;
    const double d1 = (
        std::log(spot / strike)
        + (risk_free_rate + 0.5 * variance) * time_to_maturity
    ) / (volatility * sqrt_t);
    const double d2 = d1 - volatility * sqrt_t;
    const double pdf_d1 = normal_pdf(d1);
    const double discount_factor = std::exp(-risk_free_rate * time_to_maturity);

    const double gamma = pdf_d1 / (spot * volatility * sqrt_t);
    const double vega = spot * pdf_d1 * sqrt_t;
    const double carry_term = risk_free_rate * strike * discount_factor;
    const double decay_term = -(spot * pdf_d1 * volatility) / (2.0 * sqrt_t);

    if (option_type >= 0) {
        const double delta = normal_cdf(d1);
        const double theta = decay_term - carry_term * normal_cdf(d2);
        const double rho = strike * time_to_maturity * discount_factor * normal_cdf(d2);
        return {delta, gamma, vega, theta, rho};
    }

    const double delta = normal_cdf(d1) - 1.0;
    const double theta = decay_term + carry_term * normal_cdf(-d2);
    const double rho = -strike * time_to_maturity * discount_factor * normal_cdf(-d2);
    return {delta, gamma, vega, theta, rho};
}
}  // namespace pricing

File: apps/engine/qresearch/origins/options/greeks.py
"""Black-Scholes Greeks for European vanilla options."""

from __future__ import annotations

from dataclasses import dataclass
from math import exp, pi, sqrt

try:
    from .black_scholes import OptionType, d1, d2, normal_cdf
except ImportError:  # pragma: no cover - supports direct script execution.
    from black_scholes import OptionType, d1, d2, normal_cdf

_INV_SQRT_2PI = 1.0 / sqrt(2.0 * pi)
_EPSILON = 1.0e-12


@dataclass(frozen=True)
class Greeks:
    """Option sensitivities under Black-Scholes."""

    delta: float
    gamma: float
    vega: float
    theta: float
    rho: float


def normal_pdf(x: float) -> float:
    """Standard normal density."""

    return _INV_SQRT_2PI * exp(-0.5 * x * x)


def black_scholes_greeks(
    spot: float,
    strike: float,
    time_to_maturity: float,
    risk_free_rate: float,
    volatility: float,
    option_type: OptionType = "call",
) -> Greeks:
    """Return delta, gamma, vega, theta, and rho.

    Vega is expressed per 1.00 volatility point, not per 1 percentage point.
    Theta is annualized.
    """

    if time_to_maturity <= _EPSILON or volatility <= _EPSILON:
        if option_type == "call":
            delta = 1.0 if spot > strike else 0.0 if spot < strike else 0.5
        elif option_type == "put":
            delta = -1.0 if spot < strike else 0.0 if spot > strike else -0.5
        else:
            raise ValueError(f"Unsupported option type: {option_type!r}")
        return Greeks(delta=delta, gamma=0.0, vega=0.0, theta=0.0, rho=0.0)

    d_1 = d1(spot, strike, time_to_maturity, risk_free_rate, volatility)
    d_2 = d2(spot, strike, time_to_maturity, risk_free_rate, volatility)
    pdf_d1 = normal_pdf(d_1)
    discount_factor = exp(-risk_free_rate * time_to_maturity)
    sqrt_t = sqrt(time_to_maturity)

    gamma = pdf_d1 / (spot * volatility * sqrt_t)
    vega = spot * pdf_d1 * sqrt_t

    carry_term = risk_free_rate * strike * discount_factor
    decay_term = -(spot * pdf_d1 * volatility) / (2.0 * sqrt_t)

    if option_type == "call":
        delta = normal_cdf(d_1)
        theta = decay_term - carry_term * normal_cdf(d_2)
        rho = strike * time_to_maturity * discount_factor * normal_cdf(d_2)
    elif option_type == "put":
        delta = normal_cdf(d_1) - 1.0
        theta = decay_term + carry_term * normal_cdf(-d_2)
        rho = -strike * time_to_maturity * discount_factor * normal_cdf(-d_2)
    else:
        raise ValueError(f"Unsupported option type: {option_type!r}")

    return Greeks(delta=delta, gamma=gamma, vega=vega, theta=theta, rho=rho)


def portfolio_delta(option_inventory: int, option_delta: float, multiplier: int = 100) -> float:
    """Convert contract inventory and per-option delta into share-equivalent delta."""

    return option_inventory * option_delta * multiplier
